### PR TITLE
ENH: Use custom file-like protocols instead of `typing.IO`

### PR DESCRIPTION
**Pull request by BvB93**
The problem with `typing.IO` is that it is an abstract base class, and one that demands the user to implement [a lot of methods](https://github.com/python/cpython/blob/e09dd8aafd6f9aef03945c417267806d47084a5d/Lib/typing.py#L2057-L2153). Generally speaking, file-like objects only have to implement a small subset of these methods (_e.g._ `write()`) for it to be used by a particular function, the rest quite frankly being overkill.

This PR thus replaces most usage of the much broader `typing.IO` type with smaller custom protocols, matching the convention used in typeshed (https://github.com/python/typing/discussions/829#discussioncomment-1150451).

### CODE REVIEW

File: numpy/lib/npyio.pyi [hunk 1]
```diff
@@ -50,7 +53,17 @@ _DTypeLike = Union[
...
+
+class _SupportsRead(Protocol[_CharType_co]):
+    def read(self) -> _CharType_co: ...
+
+class _SupportsReadSeek(Protocol[_CharType_co]):
+    def read(self, n: int, /) -> _CharType_co: ...
+    def seek(self, offset: int, whence: int, /) -> object: ...
+
+class _SupportsWrite(Protocol[_CharType_contra]):
+    def write(self, s: _CharType_contra, /) -> object: ...
```
Comment (BvB93) [inline numpy/lib/npyio.pyi:66]: The `write` output value is ignored by the likes of `np.save` and `np.savez`, so we can get away with annotating it as `object`. The same holds for a number of other functions, such as `np.fromfile`.

File: numpy/lib/npyio.pyi [hunk 2]
```diff
@@ -50,7 +53,17 @@ _DTypeLike = Union[
...
 class _SupportsGetItem(Protocol[_T_contra, _T_co]):
-    def __getitem__(self, key: _T_contra) -> _T_co: ...
+    def __getitem__(self, key: _T_contra, /) -> _T_co: ...
+
+class _SupportsRead(Protocol[_CharType_co]):
+    def read(self) -> _CharType_co: ...
+
+class _SupportsReadSeek(Protocol[_CharType_co]):
+    def read(self, n: int, /) -> _CharType_co: ...
+    def seek(self, offset: int, whence: int, /) -> object: ...
```
Comment (BvB93) [inline numpy/lib/npyio.pyi:63]: `np.load` always supplies the full array of optional arguments to `read` and `seek`, so they can be marked as non-optional in the protocol.

**Review by charris — APPROVED** [review]

File: numpy/lib/npyio.pyi (hunk 2 above)
Comment (BvB93) [reply numpy/lib/npyio.pyi:63]: The name is based of its counterpart in [`io`](https://docs.python.org/3/library/io.html#io.IOBase.seek) and [`typing.IO`](https://github.com/python/cpython/blob/e09dd8aafd6f9aef03945c417267806d47084a5d/Lib/typing.py#L2119-L2121).

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**